### DISCUSSION TITLE
Reading extensions

### DISCUSSION THREAD

**Discussion by testlorenzospa** [FirelyTeam/firely-net-sdk#2511] [Q&A] [answered]
Hello.

I have problems about reading Extension.

At the moment, I am reading the patient informations from this fixed bundle JSON string (it is an example)

`{ "resourceType": "Bundle", "id": "example-bundle", "type": "collection", "entry": [ { "resource": { "resourceType": "Patient", "id": "patient1", "identifier": [ { "type": { "coding": [ { "system": "http://terminology.hl7.org/CodeSystem/v2-0203", "code": "TAX", "display": "TAX ID number" } ], "text": "TAX ID number" }, "value": "ABCXYZ12L34M567N" } ], "name": [ { "given": [ "Mario" ],"family": "Rossi" } ], "gender": "male", "birthDate": "1980-01-01", "deceasedDateTime" : "2023-05-10T08:00:00Z", "maritalStatus": { "coding": [ { "system": "http://hl7.org/fhir/v3/MaritalStatus", "code": "M", "display": "Married" } ] }, "address": [ { "use": "home", "text": " Via Livorno, 20", "city": "058091", "state": "00RM", "postalCode": "00162", "country": "IT", "extension": [ { "url": "http://hl7.org/fhir/StructureDefinition/address-official", "valueBoolean": true } ] } ], "extension": [ { "url": "http://hl7.org/fhir/StructureDefinition/patient-citizenship", "valueString": "Italiana" }, { "url": "http://hl7.org/fhir/StructureDefinition/patient-birthPlace", "valueAddress": { "city": "Roma" } }, { "url": "http://hl7.org/fhir/StructureDefinition/patient-codeableBirthPlace", "valueString": "058091" }, { "url": "http://hl7.it/fhir/StructureDefinition/patient-qualification-it", "valueCodeableConcept": { "coding": [ { "system": "http://terminology.hl7.it/CodeSystem/istat-ctsi03", "code": "10000000", "display": "LICENZA ELEMENTARE" } ], "text": "LICENZA ELEMENTARE" } }, { "url": "http://hl7.it/fhir/StructureDefinition/patient-occupation-it", "valueCodeableConcept": { "coding": [ { "system": "http://terminology.hl7.it/CodeSystem/istat-professioni", "code": "1", "display": "Dirigenti ed equiparati delle università e degli enti di ricerca" } ], "text": "Dirigenti ed equiparati delle università e degli enti di ricerca"} } ] } } ] }`

For reading the Extension, I have this code

JsonSerializerOptions options = new JsonSerializerOptions().ForFhir();
Bundle bundle = JsonSerializer.Deserialize<Bundle>(jsonInput, options);

foreach (var entry in bundle.Entry)
{
    var pat_entry = (Patient)entry.Resource;
    foreach (var ext in pat_entry.Extension)
    {
        Console.WriteLine("Patient's ext URLS: " + ext.Url);
        var birth_place = ext.GetStringExtension("http://hl7.org/fhir/StructureDefinition/patient-birthPlace"); //not work
        //var birth_place = ext.GetBoolExtension("http://hl7.org/fhir/StructureDefinition/patient-birthPlace"); //not work
        //var birth_place = ext.GetExtension("http://hl7.org/fhir/StructureDefinition/patient-birthPlace"); //not work
        if (birth_place!=null) {
            Console.WriteLine("Patient's birth place: " + birth_place.Length);
        } else { Console.WriteLine("bho?");
        }
    }
}

No one instruction is working but there are Url.
Why this code not work?

Thanks.

**Comment by mbaltus** [comment]
You are not on the correct level in the tree that is your FHIR resource. Your code gets extensions out of the extension, instead of extensions of the Patient.

Use a pat_entry.GetExtension instead of your current code. Also, the GetStringExtension and GetBoolExtension will not work, since the datatype used in the birthplace extension is the Address datatype, not string or bool. This also means that you cannot use birth_place.Length. You will have to check for the City or other fields in the Address and display that.

**Comment by testlorenzospa** [comment]
Hi @mbaltus 

Thanks.

After the use of pat_entry.GetExtension, how can I get the informations about the model?
For example, as you have write birthplace is an Address model (Hl7.Fhir.Model.Address, from property Value).
But, If I assign pat_entry.GetExtension to a variable, I don't have reference for read the informations about the Address model.

`var birthXplace = pat_entry.GetExtension("http://hl7.org/fhir/StructureDefinition/patient-birthPlace");
//how read Address informations?`

Can you help me?

Many thanks.

**Reply by mbaltus** [reply, marked as answer]
With GetExtension you get the whole Extension object. Since you know that this extension's value will have an Address as datatype, you can cast the Extension.Value to its actual Address type and then use that.
Or, if you are only interested in the value anyway, you can use the GetExtensionValue method to get just the Address part:
```csharp
Address birthPlace = pat_entry.GetExtensionValue<Address>("http://hl7.org/fhir/StructureDefinition/patient-birthPlace");
```

**Comment by testlorenzospa** [comment]
Thanks for your support.

Now it works.

Thanks again.
